app/src/main/java/ru/karasevm/privatednstoggle/service/ShortcutService.kt

```diff
@@ -0,0 +1,87 @@
+package ru.karasevm.privatednstoggle.service
+
+import android.app.Service
+import android.content.Intent
+import android.os.IBinder
+import android.util.Log
+import android.widget.Toast
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.SupervisorJob
+import kotlinx.coroutines.launch
+import ru.karasevm.privatednstoggle.PrivateDNSApp
+import ru.karasevm.privatednstoggle.data.DnsServerRepository
+import ru.karasevm.privatednstoggle.util.PreferenceHelper
+import ru.karasevm.privatednstoggle.util.PrivateDNSUtils
+import ru.karasevm.privatednstoggle.util.PrivateDNSUtils.DNS_MODE_AUTO
+import ru.karasevm.privatednstoggle.util.PrivateDNSUtils.DNS_MODE_OFF
+import ru.karasevm.privatednstoggle.util.PrivateDNSUtils.DNS_MODE_PRIVATE
+import ru.karasevm.privatednstoggle.util.PrivateDNSUtils.checkForPermission
+
+class ShortcutService : Service() {
+
+    private val repository: DnsServerRepository by lazy { (application as PrivateDNSApp).repository }
+    private val job = SupervisorJob()
+    private val scope = CoroutineScope(Dispatchers.IO + job)
+
+    companion object {
+        private const val ACTION_SWITCH_MODE = "myapp://switch_mode"
+        private const val ACTION_TOGGLE_MODE = "myapp://toggle_mode"
+    }
+
+    private fun setDnsModeAndShowToast(dnsMode: String) {
+        PrivateDNSUtils.setPrivateMode(contentResolver, dnsMode)
```

Comment:
Don't set the mode to `hostname` without setting a provider or at least checking that one is set. This is undefined behavior in android, leaves the user without working DNS, and causes the app itself to crash.